klippy/webhooks.py

```diff
@@ -0,0 +1,245 @@
+# Klippy WebHooks registration and server connection
+#
+# Copyright (C) 2020 Eric Callahan <arksine.code@gmail.com>
+#
+# This file may be distributed under the terms of the GNU GPLv3 license
+import logging
+import socket
+import json
+
+AVAILABLE_METHODS = ['GET', 'POST', 'DELETE']
```

Comment:
As a very high-level comment - does the Klippy code need to be aware of GET/POST/DELETE?  Could we think of the klippy interface as just a "json api" where each request is sent without necessarily originating from http?  I that that may make the api more applicable to other users.

Reply:
Well, I was thinking the web server would require POST for all api calls.  Considering the overhead of making an api call, I'm not sure GET is ever a good match, and I can't think of a good use for DELETE.

Reply:
Generally speaking client developers will expect to use a GET request to query state and a POST request to change state.  That said, we can probably come up with a way to abstract this from Klippy.  We could enforce an endpoint naming convention.  For example:
```
webhooks.register_endpoint('query/info', cb)
webhooks.register_endpoint('action/gcode/script, cb)
```

As for DELETE, I currently use it in file_manager.py, however I can move advanced file management into the server if necessary.   All it really needs Klippy for is to make sure that a file is not currently loaded by the virtual_sdcard before deleting or overwriting it.  I could set up a "query" endpoint to do this.

Reply:
Okay.  Perhaps a little simpler would be to always accept POST api requests, and if the final component of the endpoint name starts with "query_" then allow either POST or GET - for example, `webhooks.register_endpoint('gcode/query_available_commands', self._handle_remote_help)`

Reply:
I'm ok with this.  Alternatively I could always accept POST and GET in the server's handler regardless of the name.  In the callback a WebRequestError could be raised if the method doesn't apply.  Using the `gcode/help` example:
```python
def _handle_remote_help(self, web_request):
  if web_request.get_method() != 'GET':
    raise web_request.error("Invalid Request Method")
  web_request.send(dict(self.gcode_help))
```
If the Klippy developer doesn't do this check and processes both GET and POST requests the same it would still be ok, but it would allow requests through that are technically invalid.

Reply:
Ah, I think I missed something earlier.  I thought the code was adding a json-rpc ( https://www.jsonrpc.org/specification ) interface with requests something like:
```
{"method": "gcode/script", "params": {"script": "G28"}, "id": "my-identifier"}
```
With a response something like:
```
{"result": {}, "id": "my-identifier"}
```

I see now that the method field is different.  How will the caller know when a request is complete without the tracking of the "id" field?

Reply:
Ah, I see there is a "request_id" response field.  Why the difference from the json-rpc spec, or is there some other spec I'm unaware of?

Reply:
Ah, I see.  JSON-RPC is used between the server and web clients for communication over the websocket.  Communication between Klippy and the server is a bit more simple.  

From Server To Klippy:
The server sends Klippy a  json encoded dict that a "web request" is built from:
`{'id': <unique_id>, 'path': <endpoint_path> method': <either GET or POST>, 'args': <args received with the request>}`
In this context the `method` is either `GET` or `POST`.  The path matches the endpoint registered by Klippy.  The id is a unique ID that the webserver uses to notify the co-routine waiting on a response, and args are arguments supplied by the request.

From Klippy-To-Server:
Klippy can send the following dict back to the server:
`{"method": <name of the method to execute on server>, 'params':  <keyword arguments for the method>`}
Currently available methods on the server:
- add_klippy_hook(hook)
   Configures the request handlers for the specific hook.  This will be removed now that a "web_request" will be used to
   request hooks.
- load_config(config)
   Loads server configuration parsed from printer.cfg.  This is done by the forthcoming api_server.py extra.   I'll likely remove this in favor of a "web_request" as well
- set_klippy_ready(sensors)
  Notifies the server that klipper has entered the "ready" state.  Includes available heater sensors so the server side temperature store can be initialzed (this allows clients to retrieve temperature history for each sensor when they connect)
- set_klippy_shutdown()
  Notifies the server that Klippy has entered the shutdown state
- response(request_id, response)
  Response to a web request
- notification(name, state)
  Send an async notification to clients over the websocket.  Currently this includes gcode responses and status updates.

If you would prefer to use JSON-RPC for communication between Klippy and the server we can do that.  Below is an example of how a "gcode/script" Web Request would look:

From the Server to Klippy:
`{'jsonrpc': "2.0", 'method': "gcode/script", "params": {"script": "M84"}, "id": <unique_id>}`
Klippy's Response:
`{'jsonrpc': "2.0", 'result': "ok", 'id': <id matching request>}`

In the event of an error it would look like this:
`{'jsonrpc': "2.0", 'error': {'code': <jsonrpc error code>, 'message': <message>}, 'id': <id matching request>}`
JSON-RPC has a reserved list of error codes for protocol errors (such as "Method not found", "Parse Error", etc).  It is possible for applications to define their own error codes as well, but not necessary.

Here is an example where Klippy "calls" a function on the server:
`{'jsonrpc': "2.0", 'method': "gcode_response", 'params': {'response': "// Klippy Ready"}}`

There would be no response from the server here because the "id" field is omitted.  This amounts to calling a function without a return value.  In the event of an error on the above call then Klippy would received an "error" response with the "id" field set to `None`.